syntax/grammar.mjs

```diff
@@ -0,0 +1,495 @@
+
+let Comment = defer(() =>
+    repeat1(
+        sequence(
+            char("#"),
+            comment_line.as("line")))
+    .map(flatten(1))
+    .map(to_array)
+    .map(to_string)
+    .map(into(FTL.Comment)));
+
+let GroupComment = defer(() =>
+    repeat1(
+        sequence(
+            string("##"),
+            comment_line.as("line")))
+    .map(flatten(1))
+    .map(to_array)
+    .map(to_string)
+    .map(into(FTL.GroupComment)));
+
+let ResourceComment = defer(() =>
+    repeat1(
+        sequence(
+            string("###"),
+            comment_line.as("line")))
+    .map(flatten(1))
+    .map(to_array)
+    .map(to_string)
+    .map(into(FTL.ResourceComment)));
+
+let Junk = defer(() =>
+    repeat1(
+        and(
+            not(Message),
+            not(Term),
+            not(ResourceComment),
+            not(GroupComment),
+            not(Comment),
+            sequence(
+                regex(/.*/),
+                line_end)))
+    .map(flatten(1))
+    .map(to_string)
+    .map(into(FTL.Junk)));
+
+/* ----------------------------------------------- */
+/* Patterns consist of TextElements or Placeables. */
+let Pattern = defer(() =>
+    repeat1(
+        PatternElement)
+    // Flatten indented Placeables.
+    .map(flatten(1))
+    .map(into(FTL.Pattern)));
+
+let PatternElement = defer(() =>
+    either(
+        TextElement,
+        Placeable,
+        sequence(
+            break_indent.map(into(FTL.TextElement)),
+            Placeable)));
+
+let TextElement = defer(() =>
+    repeat1(
+        either(
+            text_char,
+            text_cont))
+    .map(to_string)
+    .map(into(FTL.TextElement)));
+
+let Placeable = defer(() =>
+    sequence(
+        char("{"),
+        between(
+            maybe(inline_space),
+            either(
+                // Order matters!
+                BlockExpression,
+                InlineExpression).as("expression")),
+        char("}"))
+    .map(to_object)
+    .map(into(FTL.Placeable)));
+
+let BlockExpression = defer(() =>
+    either(
+        SelectExpression,
+        VariantList));
+
+let InlineExpression = defer(() =>
+    either(
+        StringExpression,
+        NumberExpression,
+        CallExpression, // Must be before MessageReference
+        MessageAttributeExpression,
+        TermVariantExpression,
+        Identifier.map(into(FTL.MessageReference)),
+        TermIdentifier.map(into(FTL.MessageReference)),
+        ExternalArgument,
+        Placeable));
+
+/* ------------------ */
+/* Inline Expressions */
+let StringExpression = defer(() =>
+    quoted_text.map(into(FTL.StringExpression)));
+
+let NumberExpression = defer(() =>
+    number.map(into(FTL.NumberExpression)));
+
+let ExternalArgument = defer(() =>
+    ExternalIdentifier.map(into(FTL.ExternalArgument)));
+
+let CallExpression = defer(() =>
+    sequence(
+        Function.as("callee"),
+        char("("),
+        between(
+            maybe(inline_space),
```

Comment:
I think this reads better without the `between inline_space`. We have that around each argument anyway.

Reply:
We need at least the `maybe(inline_space)` part to handle the case of no arguments: `FUNCTION( <whitespace here>     )`.